File: python/LeetCode/135_Candy.py
"""
There are N children standing in a line. Each child is assigned a rating value.

You are giving candies to these children subjected to the following requirements:

Each child must have at least one candy.
Children with a higher rating get more candies than their neighbors.
What is the minimum candies you must give?

Example 1:

Input: [1,0,2]
Output: 5
Explanation: You can allocate to the first, second and third child with 2, 1, 2 candies respectively.
Example 2:

Input: [1,2,2]
Output: 4
Explanation: You can allocate to the first, second and third child with 1, 2, 1 candies respectively.
             The third child gets 1 candy because it satisfies the above two conditions.
"""
from typing import List
class Solution:
    def candy(self, ratings: List[int]) -> int:
        
        n = len(ratings)
        cds = [1 for i in range(n)]
        for i in range(1, n):
            if ratings[i] > ratings[i-1]:
                cds[i] = cds[i-1] + 1
        for i in range(1, n):
            if ratings[-i-1] > ratings[-i]:
                cds[-i-1] = max(cds[-i] + 1,  cds[-i-1])
        return sum(cds)

"""
对这道hard难度的题目, 我觉得最重要的就是能够理解题目的意思.
其中一个容易被错误理解的点就是: 当相邻两个人rating相同时, 他们的candy是可以有差异的.
"""

File: cpp/LeetCode/135_candy.cpp
// 老师想给孩子们分发糖果，有N个孩子站成了一条直线，老师会根据每个孩子的表现，预先给他们评分。

// 你需要按照以下要求，帮助老师给这些孩子分发糖果：
// 每个孩子至少分配到 1 个糖果。
// 评分更高的孩子必须比他两侧的邻位孩子获得更多的糖果。
// 那么这样下来，老师至少需要准备多少颗糖果呢？

#include "00_commonHeader.hpp"

class Solution {
  public:
    int candy(vector<int> &ratings) {
        auto size = ratings.size();
        if (size < 2) {
            return size;
        }

        vector<int> candies(size, 1);
        for (int i = 1; i < size; i++) {
            if (ratings[i] > ratings[i - 1]) {
                candies[i] = candies[i - 1] + 1;
            }
        }
        for (int i = size - 1; i > 0; i--) {
            if (ratings[i] < ratings[i - 1]) {
                candies[i - 1] = max(candies[i - 1], candies[i] + 1);
            }
        }
        auto num_candies = accumulate(candies.begin(), candies.end(), 0);
        return num_candies;
    }
};